workflow/controller/cache_gc.go

```diff
@@ -0,0 +1,78 @@
+package controller
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	apiv1 "k8s.io/api/core/v1"
+	apierr "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+
+	"github.com/argoproj/argo-workflows/v3/util/env"
+	"github.com/argoproj/argo-workflows/v3/workflow/common"
+	controllercache "github.com/argoproj/argo-workflows/v3/workflow/controller/cache"
+	"github.com/argoproj/argo-workflows/v3/workflow/controller/indexes"
+)
+
+var gcAfterNotHitDuration = env.LookupEnvDurationOr("CACHE_GC_AFTER_NOT_HIT_DURATION", 30*time.Second)
+
+func init() {
+	log.WithField("gcAfterNotHitDuration", gcAfterNotHitDuration).Infof("Memoization caches will be garbage-collected if they have not been hit after %s", gcAfterNotHitDuration)
+}
+
+// syncAllCacheForGC syncs all cache for GC
+func (wfc *WorkflowController) syncAllCacheForGC(ctx context.Context) {
+	configMaps, err := wfc.configMapInformer.GetIndexer().ByIndex(indexes.ConfigMapLabelsIndex, common.LabelValueTypeConfigMapCache)
+	if err != nil {
+		log.WithError(err).Error("Failed to get configmaps from informer")
+	}
+
+	for _, obj := range configMaps {
+		cm, ok := obj.(*apiv1.ConfigMap)
+		if !ok {
+			log.WithField("configMap", cm.Name).Error("Unable to convert object to configmap when syncing ConfigMaps")
+			continue
+		}
+		if err := cleanupUnusedCache(ctx, wfc.kubeclientset, cm, gcAfterNotHitDuration); err != nil {
```

Comment:
* no need for global as parameter here?
* if `cleanupUnusedCache` was `wfc.cleanupUnusedCache` then it could access `wfc.kubeclientset`